filter.go

```diff
@@ -50,68 +101,328 @@ var FilterSubstringsMap = map[uint64]string{
+}
+
+func filterEncode(opType uint64, value []string) (*ber.Packet, error) {
+	var p *ber.Packet = nil
+	var err error
+
+	// condense and/or/not into one case.
+	switch opType {
+	case FilterAnd, FilterOr, FilterNot:
+		if FilterDebug {
+			fmt.Println(FilterMap[opType])
+		}
+		p = ber.Encode(ber.ClassContext, ber.TypeConstructed, ber.Tag(opType), nil, FilterMap[opType])
+	case FilterItem:
+		if FilterDebug {
+			fmt.Println("FilterItem")
+		}
+		p, err = encodeItem(value)
+	}
+	return p, err
+}
+
+func encodeItem(attrOpVal []string) (*ber.Packet, error) {
+	attr, op, val := attrOpVal[0], attrOpVal[1], attrOpVal[2]
+	if FilterDebug {
+		fmt.Println(attr, op, val)
 	}
-	return packet, nil
+
+	if op == ":=" {
+		return encodeExtensibleMatch(attr, val)
+	}
+
+	if op == "=" {
+		if val == "*" { // simple present
+			p := ber.NewString(ber.ClassContext, ber.TypePrimitive, FilterPresent, attr, FilterMap[FilterPresent])
+			return p, nil
+		} else if unescapedWildCardRegex.Match([]byte(val)) {
+			// TODO ADD escaping.
+			return encodeSubStringMatch(attr, val)
+		}
+	}
+
+	p, _ := AttributeValueAssertion(attr, op, val)
+	return p, nil
+}
+
+/*
+substrings         [4] SubstringFilter,
+
+SubstringFilter ::= SEQUENCE {
+            type    AttributeDescription,
+            -- initial and final can occur at most once
+            substrings    SEQUENCE SIZE (1..MAX) OF substring CHOICE {
+             initial        [0] AssertionValue,
+             any            [1] AssertionValue,
+             final          [2] AssertionValue } }
+*/
+
+func encodeSubStringMatch(attr, value string) (*ber.Packet, error) {
+	p := ber.Encode(ber.ClassContext, ber.TypeConstructed,
+		FilterSubstrings, nil, FilterMap[FilterSubstrings])
+	p.AppendChild(ber.NewString(ber.ClassUniversal, ber.TypePrimitive, ber.TagOctetString, attr, "type"))
+	seq := ber.Encode(ber.ClassUniversal, ber.TypeConstructed, ber.TagSequence, nil, "substrings")
+
+	pos := 0
+
+	for {
+		matches := wildCardSearchRegex.FindStringSubmatch(value[pos:])
+		if FilterDebug {
+			fmt.Println(matches)
+		}
+
+		// not match found return error
+
+		if matches == nil && pos == 0 {
+			if FilterDebug {
+				fmt.Println("Did not match filter")
+			}
+			return nil, errors.New("ldap: Did not match filter.")
+		}
+		// attr=*XXX
+		if len(matches) == 0 {
+			break
+		}
+		// initial
+		if pos == 0 && len(matches[1]) > 0 {
+			if FilterDebug {
+				fmt.Println("initial : " + matches[1])
+			}
+			seq.AppendChild(ber.NewString(ber.ClassContext, ber.TypePrimitive, FilterSubstringsInitial, UnescapeFilterValue(matches[1]), "initial"))
+		}
+		// past initial but not end
+		if pos > 0 && len(matches) > 1 && len(matches[1]) > 0 {
+			if FilterDebug {
+				fmt.Println("any : " + matches[1])
+			}
+			seq.AppendChild(ber.NewString(ber.ClassContext, ber.TypePrimitive, FilterSubstringsAny, UnescapeFilterValue(matches[1]), "any"))
+		}
+
+		pos += len(matches[0])
+		if pos == len(value) {
+			break
+		}
+	}
```

Comment:
The above for loop is what should be handling these compound substring matches. It appears that it treasts (sn=*foo*bar*) as multiple substring matches (which I think is correct), where as previously it was treated as a single one.

If my understanding is correct, perhaps the removed tests should be replaced with some tests that verify that the substring match is properly broken into multple substring segments?